src/ip_address_observer.py

```diff
@@ -0,0 +1,137 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""IP address changes observer."""
+
+import logging
+import os
+import signal
+import socket
+import subprocess
+import sys
+import time
+
+from ops.charm import CharmBase, CharmEvents
+from ops.framework import EventBase, EventSource, Object
+from ops.model import ActiveStatus
+
+logger = logging.getLogger(__name__)
+
+# File path for the spawned ip address observer process to write logs.
+LOG_FILE_PATH = "/var/log/ip_address_observer.log"
+
+
+class IPAddressChangeEvent(EventBase):
+    """A custom event for IP address change."""
+
+
+class IPAddressChangeCharmEvents(CharmEvents):
+    """A CharmEvents extension for IP address changes.
+
+    Includes :class:`IPAddressChangeEvent` in those that can be handled.
+    """
+
+    ip_address_change = EventSource(IPAddressChangeEvent)
+
+
+class IPAddressObserver(Object):
+    """Observes changes in the unit's IP address.
+
+    Observed IP address changes cause :class:`IPAddressChangeEvent` to be emitted.
+    """
+
+    def __init__(self, charm: CharmBase):
+        super().__init__(charm, "ip-address-observer")
+
+        self.charm = charm
+
+    def start_observer(self):
+        """Start the IP address observer running in a new process."""
+        if (
+            not isinstance(self.charm.unit.status, ActiveStatus)
+            or self.charm.peers is None
+            or "observer-pid" in self.charm.unit_peer_data
+        ):
+            return
+
+        logger.info("Starting IP address observer process")
+
+        # We need to trick Juju into thinking that we are not running
+        # in a hook context, as Juju will disallow use of juju-run.
+        new_env = os.environ.copy()
+        if "JUJU_CONTEXT_ID" in new_env:
+            new_env.pop("JUJU_CONTEXT_ID")
+
+        pid = subprocess.Popen(
+            [
+                "/usr/bin/python3",
+                "src/ip_address_observer.py",
```

Comment:
I believe the `venv` folder should be on PYTHONPATH (since PYTHONPATH is a copy of the `os.environ`). I have tested that this approach works - we can modify it in the future if it stops working